crates/resolver/src/name_server/name_server.rs

```diff
@@ -117,104 +116,157 @@ impl<P: ConnectionProvider> Debug for NameServer<P> {
 }
 
 struct NameServerState<P: ConnectionProvider> {
-    ip: IpAddr,
-    config: ConnectionConfig,
+    config: NameServerConfig,
     options: Arc<ResolverOpts>,
     tls: Arc<TlsConfig>,
-    client: AsyncMutex<Option<P::Conn>>,
-    meta: ConnectionMeta,
-    trust_negative_responses: bool,
+    connections: AsyncMutex<Vec<ConnectionState<P>>>,
+    server_srtt: DecayingSrtt,
     connection_provider: P,
 }
 
 impl<P: ConnectionProvider> NameServerState<P> {
     fn new(
-        server_config: &NameServerConfig,
-        config: ConnectionConfig,
+        connections: Vec<ConnectionState<P>>,
+        config: NameServerConfig,
         options: Arc<ResolverOpts>,
         tls: Arc<TlsConfig>,
-        client: Option<P::Conn>,
         connection_provider: P,
     ) -> Self {
         Self {
-            ip: server_config.ip,
             config,
             options,
             tls,
-            client: AsyncMutex::new(client),
-            meta: ConnectionMeta::default(),
-            trust_negative_responses: server_config.trust_negative_responses,
+            connections: AsyncMutex::new(connections),
+            server_srtt: DecayingSrtt::new(Duration::from_micros(rand::random_range(1..32))),
             connection_provider,
         }
     }
 
-    async fn send(self: Arc<Self>, request: DnsRequest) -> Result<DnsResponse, ProtoError> {
-        let mut prev = self.client.lock().await;
-
-        // if this is in a failure state
-        let client = match (&*prev, self.status()) {
-            // if the client is already connected, then use it
-            (Some(client), Status::Init | Status::Established) => {
-                debug!(config = ?self.config, "using existing connection");
-                client.clone()
-            }
-            _ => {
-                self.set_status(Status::Init);
-                debug!(config = ?self.config, "connecting");
-
-                let new_client = Box::pin(self.connection_provider.new_connection(
-                    self.ip,
-                    &self.config,
-                    &self.options,
-                    &self.tls,
-                )?)
-                .await?;
-
-                // establish a new connection
-                *prev = Some(new_client.clone());
-                new_client
-            }
+    async fn send(
+        self: Arc<Self>,
+        request: DnsRequest,
+        skip_udp: bool,
+    ) -> Result<DnsResponse, ProtoError> {
+        let mut connections = self.connections.lock().await;
+        connections.retain(|conn| matches!(conn.meta.status(), Status::Init | Status::Established));
+        let (conn, meta) = if connections.is_empty() {
```

Comment:
It feels like maybe `connections` could stand to be its own type that encapsulates some of the work required to filter to produce a new connection to use (either making one if the contents are empty, or sorting the existing ones to produce the best-fitting pre-existing one). 

That would let this method focus more on the job to be done with a connection in-hand. WDYT?